src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+        return Selection::fromValues(values_.begin(), values_.end());
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", values_);
+    }
+
+  private:
+    Selection::Values values_;
+};
+
+using CompoundTargets = std::vector<std::string>;
+class NodeSetCompoundRule: public NodeSetRule
+{
+  public:
+    NodeSetCompoundRule(std::string attribute, const CompoundTargets& targets)
+        : attribute_(std::move(attribute))
+        , targets_(targets) {}
+
+    Selection materialize(const NodeSets& ns, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& target : targets_) {
+            ret = ret | ns.materialize(target, np);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", targets_);
+    }
+
+  private:
+    std::string attribute_;
+    CompoundTargets targets_;
+};
+
+int64_t get_integer_or_throw(const json& el) {
+    auto v = el.get<double>();
+    if (std::floor(v) != v) {
+        throw SonataError("Only allowed integers in node set rules");
+    }
+    return static_cast<int64_t>(v);
+}
+
+NodeSetRules _dispatch_node(const json& contents) {
+    NodeSetRules ret;
+
+    for (auto& el : contents.items()) {
+        const std::string& attribute = el.key();
+        if (el.value().is_number()) {
+            if (attribute == "population") {
+                throw SonataError("'population' must be a string");
+            }
+
+            int64_t v = get_integer_or_throw(el.value());
+            if (attribute == "node_id") {
+                if (v < 0) {
+                    throw SonataError("'node_id' must be positive");
+                }
+
+                ret.emplace_back(new NodeSetBasicNodeIds({static_cast<Selection::Value>(v)}));
+            } else {
+                ret.emplace_back(new NodeSetBasicRule<int64_t>(attribute, {v}));
+            }
+        } else if (el.value().is_string()) {
+            if (attribute == "node_id") {
+                throw SonataError("'node_id' must be numeric or a list of numbers");
+            }
+
+            if (attribute == "population") {
+                ret.emplace_back(new NodeSetBasicPopulation(el.value().get<std::string>()));
+            } else {
+                ret.emplace_back(
+                    new NodeSetBasicRule<std::string>(attribute, {el.value().get<std::string>()}));
+            }
+        } else if (el.value().is_array()) {
+            const auto array = el.value();
+
+            if (attribute == "population" && (array.size() != 1 || !array[0].is_string())) {
+                throw SonataError("'population' must be a single string");
+            }
+
+            if (array[0].is_number()) {
+                std::vector<int64_t> values;
+                for (auto& inner_el : array.items()) {
+                    values.emplace_back(get_integer_or_throw(inner_el.value()));
```

Comment:
for int is there a benefit in using emplace_back instead of push_back ?

Reply:
Not in this case; but I'm using `emplace_back` elsewhere, and there's no cost, afaik.

Reply:
ok it was for my personal knowledge thx !